### PR TITLE
Python implementation of pgMAP

### PR DESCRIPTION
**Pull request by marissafujimoto**
# Description

 

This is the implementation of the one shot pgMAP counter algorithm in python.
image

## Type of change

 

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

 

- [x] Unit and integration tests written with python unittest package. Run with 
```console
python3 -m pip install . && python3 -m tests
```

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas (IN PROGRESS)
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+[build-system]
```
Comment (marissafujimoto) [inline pyproject.toml:1]: Just added this as a basic build system, no real preference beyond that

File: tests/__main__.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+from pgmap import counter
+from pgmap.model.paired_read import PairedRead
+
+
+class TestPgmap(unittest.TestCase):
+
+    def test_read_fastq(self):
+        count = 0
+
+        for sequence in fastx_reader.read_fastq("example-data/three-read-strategy/HeLa/PP_pgRNA_HeLa_S1_I1_001_Sampled10k.fastq.gz"):
```
Comment (marissafujimoto) [inline tests/__main__.py:15]: Just saw this is a copy paste error. Should test on a non .gz file. Will fix in a later commit

File: tests/__main__.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+        gRNA_mappings = {"LET": {"WOW", "LEG", "EAT"}}
+        candidate_reads = [PairedRead("LET", "ROT", "FOOD"),
+                           PairedRead("LET", "EXT", "FXOD"),
+                           PairedRead("RUN", "LEG", "WOOD")]
+
+        paired_guide_counts = counter.get_counts(
+            candidate_reads, gRNA_mappings, barcodes)
+
+        self.assertEqual(sum(paired_guide_counts.values()), 2)
+
```
Comment (marissafujimoto) [inline tests/__main__.py:172]: TODO: verify key, values of paired_guide_counts match as expected.

**Review by cansavvy — APPROVED** [review]
@marissafujimoto I think this overall setup looks good from what I can tell so far! :tada: Thanks for getting this going! I would just say before we dive too much more in the weeds we may want to draft a document or mock code example that demonstrates what you envision the usage of this to look like. 

So an example usage that we can use to guide our development. This example should include all the steps that we think a user is going to do in the context of pgmap. 

The overall steps of this example should probably include something like: 
1. Read in a pgRNA library 
2. Do some kind of quality control check
3. Point to their fastq files to actually do the counting
4. (Optional) Maybe make a nice plot or table proof of concept purposes

And for each point we'll want to think about what parameters and arguments will we want to be available for users to specify. 

If we think about this first then it will help us discuss the details of the code you have started here.

**Comment by cansavvy** [comment]
```
ssh [Hutch]@rhino
**Put in password**
```
Set up Jupyter
```
ml purge
ml JupyterLab/4.0.3-GCCcore-12.2.0  Seaborn/0.12.2-foss-2022b scikit-learn/1.2.1-gfbf-2022b
jupyter lab --ip=$(hostname) --port=$(fhfreeport) --no-browser
```

In terminal in Jupyter: 
```
git clone https://github.com/FredHutch/pgmap
cd pgmap
git checkout python-implementation
```
Checkout other branch
```
cd ../pg-notebooks/
git branch pipeline-comparison
```
In notebook:
```
!python3 -m pip install ../pgmap editdistance
```

**Review by cansavvy — APPROVED** [review]
I think this is a great start!

**Pull request merged**